File: src/main/java/dev/obscuria/obscure_tooltips/client/registry/ResourceRegistry.java
package dev.obscuria.obscure_tooltips.client.registry;

import com.google.common.collect.Lists;
import net.minecraft.util.ResourceLocation;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public class ResourceRegistry<T> {
    protected final String name;
    protected final Map<ResourceLocation, T> keyToElement = new ConcurrentHashMap<>();
    protected final Map<T, ResourceLocation> elementToKey = new ConcurrentHashMap<>();

    public ResourceRegistry(String name) {
        this.name = name;
    }

    public void register(ResourceLocation key, T element) {
        keyToElement.put(key, element);
        elementToKey.put(element, key);
    }

    public Collection<T> listElements() {
        return keyToElement.values();
    }

    public T get(ResourceLocation key) {
        return keyToElement.get(key);
    }

    public void onReloadStart() {
        keyToElement.clear();
        elementToKey.clear();
    }

    public void onReloadEnd() {}

    public int total() {
        return keyToElement.size();
    }

    @Override
    public String toString() {
        return "ResourceRegistry[" + this.name + "]";
    }


    public static class Ordered<T extends Comparable<T>> extends ResourceRegistry<T> {
        private final List<T> sortedElements = Lists.newArrayList();

        public Ordered(String name) {
            super(name);
        }

        @Override
        public Collection<T> listElements() {
            return sortedElements;
        }

        @Override
        public void onReloadEnd() {
            sortedElements.clear();
            sortedElements.addAll(keyToElement.values());
            sortedElements.sort(Comparator.reverseOrder());
        }
    }
}

File: common/src/main/kotlin/dev/obscuria/tooltips/registry/ResourceRegistry.kt
package dev.obscuria.tooltips.registry

import com.mojang.serialization.Codec
import com.mojang.serialization.DataResult
import net.minecraft.resources.ResourceLocation
import java.util.*

open class ResourceRegistry<T> {

    internal val elementByKey: MutableMap<ResourceLocation, T> = mutableMapOf()
    internal val keyByElement: MutableMap<T, ResourceLocation> = mutableMapOf()

    val total: Int get() = elementByKey.size

    fun register(key: ResourceLocation, element: T) {
        elementByKey[key] = element
        keyByElement[element] = key
    }

    fun byNameCodec(): Codec<T> {
        return ResourceLocation.CODEC.flatXmap(::tryGetElement, ::tryGetKey)
    }

    open fun listElements(): Collection<T> {
        return elementByKey.values
    }

    open fun onReloadStart() {
        elementByKey.clear()
    }

    open fun onReloadEnd() {}

    private fun tryGetElement(key: ResourceLocation): DataResult<T> {
        return Optional.ofNullable(elementByKey[key])
            .map { DataResult.success(it) }
            .orElseGet { DataResult.error<T> { "Unknown registry key in ${javaClass.simpleName}: $key" } }
    }

    private fun tryGetKey(element: T): DataResult<ResourceLocation> {
        return Optional.ofNullable(keyByElement[element])
            .map { DataResult.success(it) }
            .orElseGet { DataResult.error { "Unknown registry element in ${javaClass.simpleName}: $element" } }
    }

    class Ordered<T> : ResourceRegistry<T>() where T : Comparable<T> {

        private var sortedElements: List<T> = mutableListOf()

        override fun listElements(): Collection<T> {
            return sortedElements
        }

        override fun onReloadEnd() {
            sortedElements = elementByKey.values.sortedDescending()
        }
    }
}